### DISCUSSION TITLE
Can't convert Channel to List to use Groovy methods in DSL2

### DISCUSSION THREAD

**Discussion by chbk** [nextflow-io/nextflow#2547] [Q&A] [answered]
## Bug report 

Similar to #2274 but I haven't found a satisfying solution. Using DSL2, it seems Channels can no longer be converted to Lists.
I need a List to use Groovy methods.

DSL1 ✅ 

```Nextflow
list = Channel.of(1, 2, 3).toList().get()
println(list.toString()) // outputs [1, 2, 3]
```

DSL2 :x:

```Nextflow
nextflow.enable.dsl = 2
list = Channel.of(1, 2, 3).toList().get() // freezes
println(list.toString())
```

DSL1 ✅ 

```Nextflow
list = []
Channel.of(1, 2, 3).toList().subscribe(onNext: { it -> list = it })
println(list.toString()) // outputs [1, 2, 3]
```

DSL2 :x:

```Nextflow
nextflow.enable.dsl = 2
list = []
Channel.of(1, 2, 3).toList().subscribe(onNext: { it -> list = it })
println(list.toString()) // outputs []
```

Nextflow version: 21.04.3

**Comment by pditommaso** [comment, marked as answer]
The use of `.get` does not make part of Nextflow DSL and it's not expected to work with DSL2. The last example is correct due to the asynchronous nature of the nextflow channels.

**Reply by chbk** [reply]
Alright thank you for the clarification.

**Reply by hmassalha** [reply]
Unfortunately the Nextflow just confuses the uses with DSL 2 and 1. the tutorials become too long and with useless information since you never can predict what will work with dsl1 or 2 + new error messages!

**Reply by mribeirodantas** [reply]
Most of the time, the issues I see are when people try to write "pure Groovy" code to operate with things that exist only in the Nextflow DSL. Processes and operators are not regular functions. Channels are not regular lists/collections. If you take this into consideration, and try to stick to the Nextflow DSL to interact with things brought up by the Nextflow DSL, you should be fine.

As a good source for migration, I suggest checking out [this link](https://www.nextflow.io/docs/latest/dsl2.html).

**Comment by ebioman** [comment]
Quick question as I am struggling for an onComplete function which I used for DSL1 and does not work anymore work DSL2.
I used a channel and did a `for element in MyChannel` grovvy function to loop over them and do some after checks.
Now, similarly I am facing the issue that one can not easily with groovy access channels similar to a list - so is there a different way to : 
- convert them to a list
- access the elements of the object ?

**Reply by mribeirodantas** [reply]
You can use the [`collect`](https://www.nextflow.io/docs/latest/operator.html#collect) channel operator to convert a channel into a list, and use a closure to do something on the elements of this list. Check the example below:
```
Channel
    .of(1, 2, 3, 4)
    // Check if elements are odd numbers
    // and convert to zero if so. Otherwise
    // keep the elements as they are.
    .collect { it % 2 ? 0 : it }
    .view()
```
The output is:
```
[0, 2, 0, 4]
```

You can also use the [`branch`](https://www.nextflow.io/docs/latest/operator.html?highlight=subscribe#branch) channel operator to call different processes depending on a check on the channel elements. For example:
```
def criteria = branchCriteria {
    small: it < 10
    large: it > 10
}

Channel.of(1, 2, 30).branch(criteria).set { ch1 }
Channel.of(10, 20, 1).branch(criteria).set { ch2 }
```
Then you can call `processA` witht `ch1` and `processB` with `ch2`, for example.

You can also use the [`subscribe`](https://www.nextflow.io/docs/latest/channel.html?highlight=subscribe#subscribe) method and do something like:
```
Channel
    .of(1, 1, 2, 2, 2, 3, 1, 1, 2, 2, 3)
    .distinct()
    .subscribe onNext: { println it }, onComplete: { println 'Done' }
```
The output is:
```
1
2
3
1
2
3
Done
```

My suggestion is to always look for Nextflow channel operators before considering write some custom Groovy code to operate on channels 😃

**Comment by ebioman** [comment]
Thanks, completely understand and agree as well. As it is though for the
unstaging in the onComplete function, groovy has there it's justification
in my eyes.
Would never use it though indeed for anything within processes or to shape
channels. This should indeed be discouraged 😉
